text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,182 @@
+run-image-mirrors = ["scl.sh/run"]
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[extend.build] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+```
+
+Behavior: creates a new version of an existing builder with additional packages as well as a new run image (`sclevine/run`) with additional packages.
+
+If `extend.run-image` is not specified, `extend.run` is stored on the new builder metadata and used to dynamically extend the run image on `pack build`.
+
+Note that the `[builder]` section is mutually exclusive with `[stack]`.
+
+### Building an App with Additional Packages
+
+`pack build sclevine/myapp`
+
+project.toml:
+```toml
+...
+[run.extend] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[build.extend] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+...
+```
+
+Behavior: creates a version of the current builder with additional packages and an ephemeral run image with additional packages, then does a normal `pack build`.
+
+Question: should we store the new builder image to make rebuild faster? If so, where? Should we generate a tag for it?
+
+### Upgrade an Extended Builder
+
+Both build and run images: `pack upgrade sclevine/builder`
+
+Run images (including mirrors): `pack upgrade sclevine/run`
+
+### Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Specification
+
+Platforms may extend app images using the above interface by running the `extend` executable as root in a new container and creating an image from the result.
+Extending an image should generate a single layer.
+This should happen prior to the normal CNB build or rebase process.
+The `*-metadata-toml-file` file is used to determine what packages to install.
+The `pkg-cache-dir` directory is used to cache package databases (e.g. to speed up `apt-get update`).
+
+## UX
+
+`pack upgrade` (on base image)
+1. See if new base image available, if so proceed to (3). Otherwise, (2).
+2. Run status, if updates available, proceed to (3). Otherwise, stop.
+3. Run extend and publish new image. Store extension metadata on images. 
+
+If running `pack upgrade` on a builder, perform operation against both builder and run images and publish both.
+If running `pack upgrade` on an app, perform operation against run image (with ephemeral output into app repo), then `pack rebase`.  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Requires kaniko (or a similar tool) when a docker daemon is unavailable
+- Fully rebasing apps with package dependencies becomes many orders of magnitude slower, due to upgrade needs
```

Comment:
Can this scenario be drawn out more? I'm trying to understand the exact workflow for an app maintainer to rebase their app images.

Wouldn't it be as simple as:

1. I need to update my run images: `pack upgrade my/run-image` - takes some time to install packages
2. I now rebase my app images: `pack rebase my/app` - near-instant

I don't know that the app image cares that it comes from this "extended" run image.

Reply:
Just meant that the full rebase process (either with the two commands you specified, or with a single `pack upgrade my/app`) takes longer than having stack authors include more packages in stacks (or having a larger selection of dependency-providing buildpacks).